### ISSUE TITLE
Strange 3d reconstruction result

### ISSUE THREAD

**Issue by zlukich** [Anttwo/SuGaR#123]
Hi, I am trying to use your project to create verification station for 3D printed parts, to measure how precise object printed I compare point clouds of reconstructed object and ground truth point cloud.
After training on several objects with different amount of training images and different parameters I get some cursed refined mesh output even though colmap scene is perfect. Also same setup using instant-ngp gives perfect result, so I am trying to understand where is the problem here.

![colmap output](https://github.com/Anttwo/SuGaR/assets/48150694/8dd481d6-1a71-42e1-b9af-19896d86db09)

![sugar output](https://github.com/Anttwo/SuGaR/assets/48150694/1687af51-ee7d-4680-9e4b-b06d4c648abb)

As I understand standard output of SuGar .obj file from refined mesh represents Gaussian splats and not mesh in usual understanding, isn't it? Can problem lay in way how I open this file? 

Thanks

**Comment by Anttwo** [comment]
Hello @zlukich,

Indeed, your resulting mesh looks quite bad.
The output `.OBJ` file is a mesh, so you should get a good-looking result if you open it with Blender, for example.
SuGaR should be able to produce better meshes/reconstructions than methods like Instant-NGP or NeuS.

So, let's investigate!
May I ask a few questions:
1. Could you show me a few images of your dataset?
2. Did you segment the object in your input images? Currently, there's not mask compatibility implemented in SuGaR (but I'm planning to add this functionality in near future). Therefore, if you use masked images (with a monochrome, masked background), you will indeed get some chaotic geometry in the background. That could explain this cursed result. If so, you can try using the unmasked images.

**Comment by zlukich** [comment]
I think I need also explain a bit, that don't move camera, my object rotates at rotating bed, while camera stays static. We tested it on several objects and scene configurations, and ended with ambient lighting from above and static background(black mate surface). With other reconstruction methods, it was easy to work when we removed background, so I tried it in your method.

1. In this specific case I used images with removed background (73 images), because my target is to get object without any scene background information, but after you said that, I think problem can lay here (in INGP it works though, mb it is somehow possible to run without background in Sugar?)

![0007](https://github.com/Anttwo/SuGaR/assets/48150694/b1d05992-97d5-424e-bd8a-93dbe82cf5ee)
![0021](https://github.com/Anttwo/SuGaR/assets/48150694/f690bf41-f5e8-4ca6-afb6-0eed0f85e429)

2.  This is result with background
<img width="697" alt="frog_video" src="https://github.com/Anttwo/SuGaR/assets/48150694/93fe7e82-5b9b-4a2c-b99f-cf00a6b8e3f8">
<img width="670" alt="006 jpg" src="https://github.com/Anttwo/SuGaR/assets/48150694/ca620f43-9aba-4d92-a823-cec91dc347ea">

3.  Ambient light is already in use, intensity can be modified 

So to my intuition, scene can be Improved with lightning and capture better image quality, but also scene setup plays important role, because flat surface is corrupted, and when we try to remove background it optimizes on dark pixels somehow wrong. I would be glad, if you share your opinion on that, thanks :)

**Comment by Anttwo** [comment]
You're right, the setup is important; Here, your images are close-ups on the main object, placed on a monochrome floor. 
Unfortunately, a flat, monochrome surface like this generally results in chaotic reconstructions when using photogrammetry/Neural surface methods/SuGaR because there is too much ambiguity in the projective inverse problem (from a maths point of view, a chaotic geometry on the floor is a valid solution to solve this projective inverse problem).

**I agree that in this case, using masks is the best way to solve the problem; Unfortunately, there's no mask functionality for SuGaR currently.**
I really should implement a mask functionality as soon as possible. It's currently the top priority on my to-do list for this repo.

In the meantime, you could try to capture your object on a textured floor (you can add a piece of cloth or paper with texture/drawings on it), and you should be able to reconstruct the object without having such chaotic geometry.

_**Concerning Instant-NGP**: when you say that the reconstruction looks good, do you mean the images rendered by the NeRF (since Instant-NGP does not provide a mesh)? Actually, achieving good-looking rendered images is easier than reconstructing a nice surface. For instance, if you visualize the hybrid representation of the SuGaR reconstruction rather than just the mesh, the result would probably be nice too. You can use our dedicated viewer to visualize the hybrid representation._

**Comment by zlukich** [comment]
> You're right, the setup is important; Here, your images are close-ups on the main object, placed on a monochrome floor. Unfortunately, a flat, monochrome surface like this generally results in chaotic reconstructions when using photogrammetry/Neural surface methods/SuGaR because there is too much ambiguity in the projective inverse problem (from a maths point of view, a chaotic geometry on the floor is a valid solution to solve this projective inverse problem).
> 
> **I agree that in this case, using masks is the best way to solve the problem; Unfortunately, there's no mask functionality for SuGaR currently, I'm sorry for that.** I really should implement a mask functionality as soon as possible. It's currently the top priority on my to-do list for this repo.
> 
> In the meantime, you could try to capture your object on a textured floor (you can add a piece of cloth or paper with texture/drawings on it), and you should be able to reconstruct the object without having such chaotic geometry.
> 
> _**Concerning Instant-NGP**: when you say that the reconstruction looks good, do you mean the images rendered by the NeRF (since Instant-NGP does not provide a mesh)? Actually, achieving good-looking rendered images is easier than reconstructing a nice surface. For instance, if you visualize the hybrid representation of the SuGaR reconstruction rather than just the mesh, the result would probably be nice too. You can use our dedicated viewer to visualize the hybrid representation._

No when I compare results with INGP I compare them taking in accountant mesh reconstruction(previous comparisons were built on sugar on masked dataasets, so it was incorrect assumption). I retrained model with flat surface but with different colorful ground, and here you can see results from sugar and ingp.

<img width="512" alt="frog_chess_ingp_cropped" src="https://github.com/Anttwo/SuGaR/assets/48150694/86cfd068-adc1-4f0f-b521-5f12fbfad6a4" title="INGP reconstructed mesh">
<img width="569" alt="frog_chess_ingp_cropped" src="https://github.com/Anttwo/SuGaR/assets/48150694/220b0856-4177-4409-a59a-2ffe451e7d2a" title="SUGARreconstructed mesh">

From the images we also see that number of vertices and faces from INGP mush higher because we ran marching cubes algorithm with high resolution, but is there available for Sugar some kind of parameters to improve mesh quality, or adjust mesh reconstruction?
